pact/Pact/Core/IR/Eval/CEK.hs

```diff
@@ -731,30 +730,41 @@ acquireModuleAdmin i cont handler env mdl = do
    -> Eval (EvalResult CEKBigStep CoreBuiltin () Eval)
     #-}
 
+-- | Evaluate a term with all the stack manipulation logic
 evalWithStackFrame
   :: (CEKEval step b i m, MonadEval b i m)
   => i
   -> Cont step b i m
   -> CEKErrorHandler step b i m
   -> CEKEnv step b i m
-  -> StackFrame
   -> Maybe Type
+  -> StackFrame i
   -> EvalTerm b i
   -> m (CEKEvalResult step b i m)
-evalWithStackFrame info cont handler env sf mty body = do
+evalWithStackFrame info cont handler env mty sf body = do
   cont' <- pushStackFrame info cont mty sf
   evalCEK cont' handler env body
 
+-- | Push a stack frame into the stack, and check it for recursion
 pushStackFrame
   :: (MonadEval b i m)
   => i
   -> Cont step b i m
   -> Maybe Type
-  -> StackFrame
+  -> StackFrame i
   -> m (Cont step b i m)
 pushStackFrame info cont mty sf = do
+  checkRecursion
   esStack %== (sf :)
   pure (StackPopC info mty cont)
+  where
+  checkRecursion = do
+    RecursionCheck currentCalled <- usesEvalState esCheckRecursion NE.head
+    let qn = fqnToQualName (_sfName sf)
+    when (S.member qn currentCalled) $ throwExecutionError info (RuntimeRecursionDetected qn)
```

Comment:
At first I was a little wary that we forbid _any_ kind of mutual recursion, but then I figured that we're already going to great lengths to forbid all kind of possibly well-founded (but possibly not!) recursion cases, so that's OK.

Reply:
👍 

This is just reinstating prod behavior, just slightly more efficiently using sets instead of re-traversing the callstack.